### PR TITLE
[TACHYON-1304] Port WorkerClient to use ClientBase

### PR DESCRIPTION
**Pull request by shuhuai007**
Please review
https://tachyon.atlassian.net/browse/TACHYON-1304

### CODE REVIEW

File: clients/unshaded/src/main/java/tachyon/client/block/BlockWorkerClientPool.java [hunk 1]
```diff
@@ -59,7 +57,7 @@ public void release(WorkerClient workerClient) {
     try {
       // Heartbeat to send the client metrics.
       workerClient.sessionHeartbeat();
-    } catch (IOException ioe) {
+    } catch (Exception ioe) {
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/block/BlockWorkerClientPool.java]: NIT: you may want to change variable name to `e` to match the type.

File: common/src/main/java/tachyon/ClientBase.java [hunk 2]
```diff
@@ -189,15 +189,22 @@ public synchronized void disconnect() {
...
-        mProtocol.getTransport().close();
-      }
+      closeOperation();
     } finally {
       afterDisconnect();
     }
   }
 
   /**
+   * Close the connection.
```
Comment (yupeng9) [inline common/src/main/java/tachyon/ClientBase.java]: Close -> Closes

File: common/src/main/java/tachyon/ClientBase.java [hunk 3]
```diff
@@ -189,15 +189,22 @@ public synchronized void disconnect() {
...
+   */
+  protected void closeOperation() {
```
Comment (yupeng9) [inline common/src/main/java/tachyon/ClientBase.java]: Rename this to `closeTransport` to be more specific?

File: common/src/main/java/tachyon/util/network/NetworkAddressUtils.java [hunk 4]
```diff
@@ -543,4 +543,38 @@ public static InetSocketAddress parseInetSocketAddress(String address) throws IO
     }
     return new InetSocketAddress(strArr[0], Integer.parseInt(strArr[1]));
   }
+
+  /**
+   * Extract rpcPort InetSocketAddress from Tachyon representation of network address.
```
Comment (yupeng9) [inline common/src/main/java/tachyon/util/network/NetworkAddressUtils.java]: Extract -> Extracts

File: common/src/main/java/tachyon/util/network/NetworkAddressUtils.java [hunk 5]
```diff
@@ -543,4 +543,38 @@ public static InetSocketAddress parseInetSocketAddress(String address) throws IO
...
+      String host = getFqdnHost(netAddress);
+      int port = netAddress.rpcPort;
+      return new InetSocketAddress(host, port);
+    } catch (UnknownHostException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Extract dataPort InetSocketAddress from Tachyon representation of network address.
```
Comment (yupeng9) [inline common/src/main/java/tachyon/util/network/NetworkAddressUtils.java]: ditto

**Comment by yupeng9** [comment]
Thanks @shuhuai007 for this refactoring! The worker client looks much better afterwards. The PR overall looks good to me, and I left a few minor comments.

**Comment by apc999** [comment]
https://github.com/amplab/tachyon/pull/2170 moves this `WorkerClient` from `common` to `clients`. If that PR goes in first, you may need to resolve merge conflict (which should be fine as that PR has no real code change)

File: common/src/main/java/tachyon/worker/WorkerClient.java [hunk 6]
```diff
@@ -78,35 +71,36 @@
...
-    mWorkerNetAddress = Preconditions.checkNotNull(workerNetAddress);
+    super(NetworkAddressUtils.getRpcPortSocketAddress(workerNetAddress), conf, "worker");
+    mWorkerDataServerAddress = NetworkAddressUtils.getDataPortSocketAddress(workerNetAddress);
     mExecutorService = Preconditions.checkNotNull(executorService);
-    mTachyonConf = Preconditions.checkNotNull(conf);
     mSessionId = sessionId;
     mIsLocal = isLocal;
     mClientMetrics = Preconditions.checkNotNull(clientMetrics);
     mHeartbeatExecutor = new WorkerClientHeartbeatExecutor(this);
+
```
Comment (apc999) [inline common/src/main/java/tachyon/worker/WorkerClient.java]: NIT: unnecessary empty line

File: common/src/main/java/tachyon/ClientBase.java (hunk 3 above)
Comment (apc999) [inline common/src/main/java/tachyon/ClientBase.java]: How about still have
```java
if (mProtocol != null) {
   mProtocol.getTransport().close();
}```
inside `disconnect`, but before that piece of code, calling a method named `beforeDisconnect`.  `beforeDisconnect` is an empty method in `ClientBase`, but meant to be a hook method for subclasses, just like `afterConnect`. I feel this approach might be a bit simpler and easier to understand.

File: common/src/main/java/tachyon/worker/WorkerClient.java [hunk 7]
```diff
@@ -114,101 +108,106 @@ public synchronized void accessBlock(long blockId) throws IOException {
...
-        }
-        mProtocol.getTransport().close();
-      } finally {
-        if (mHeartbeat != null) {
-          mHeartbeat.cancel(true);
-        }
-      }
-      mConnected = false;
+  protected void closeOperation() {
+    // Heartbeat to send the client metrics.
```
Comment (apc999) [inline common/src/main/java/tachyon/worker/WorkerClient.java]: As I mentioned in `ClientBase`, if we override `beforeDisconnect` here, we don't need to call method from `super`.

File: common/src/main/java/tachyon/worker/WorkerClient.java [hunk 8]
```diff
@@ -299,10 +289,12 @@ public synchronized String lockBlock(long blockId) throws IOException {
    *
    * @throws IOException
    */
-  public synchronized void mustConnect() throws IOException {
+  @Override
+  public synchronized void connect() throws IOException {
```
Comment (apc999) [inline common/src/main/java/tachyon/worker/WorkerClient.java]: This  `connect` is so different from the `connect` method in superclass. Actually the `super` `connect` method is meant to be general enough for different implementations. Could you leave a TODO here and reminding us in followup PRs to merge the connect logic in this method into the super class.

**Comment by apc999** [comment]
Thanks for the improvement! I left a few comments.

File: clients/unshaded/src/main/java/tachyon/client/block/BlockWorkerClientPool.java (hunk 1 above)
Comment (shuhuai007) [inline clients/unshaded/src/main/java/tachyon/client/block/BlockWorkerClientPool.java]: done

File: common/src/main/java/tachyon/ClientBase.java (hunk 3 above)
Comment (shuhuai007) [inline common/src/main/java/tachyon/ClientBase.java]: done

File: common/src/main/java/tachyon/util/network/NetworkAddressUtils.java (hunk 4 above)
Comment (shuhuai007) [inline common/src/main/java/tachyon/util/network/NetworkAddressUtils.java]: done

File: common/src/main/java/tachyon/util/network/NetworkAddressUtils.java (hunk 5 above)
Comment (shuhuai007) [inline common/src/main/java/tachyon/util/network/NetworkAddressUtils.java]: done

File: common/src/main/java/tachyon/worker/WorkerClient.java (hunk 8 above)
Comment (shuhuai007) [inline common/src/main/java/tachyon/worker/WorkerClient.java]: done

**Comment by aaudiber** [comment]
LGTM, thanks for the improvement!

**Comment by yupeng9** [comment]
LGTM

**Comment by yupeng9** [comment]
@apc999, are all your comments addressed for this PR?

**Comment by apc999** [comment]
+1 LGTM

**Pull request merged**